System.Data.SqlClient in Cake.CoreCLR Assembly Load Error

### What You Are Seeing?
When referencing System.Data.SqlClient in my Cake file, I get the following exception when run from the global tool in a Docker container:

```
Assembly with same name is already loaded
```
When running from Windows (`build.ps1`), the reference isn't needed, and even when present, causes no issues.

I asked for assistance in the chat, and was directed here.

### What is Expected?
The Cake script should compile.

### What version of Cake are you using?
dotnet core global tool 0.30.0

### Are you running on a 32 or 64 bit system?
64-bit

### What environment are you running on?  Windows? Linux? Mac?
Docker (Linux on Windows) `microsoft/dotnet:2.1.400-sdk-stretch`

### Are you running on a CI Server?  If so, which one?
No

### How Did You Get This To Happen? (Steps to Reproduce)
In my `build.cake` file:
```
#addin "nuget:?package=System.Data.SqlClient&version=4.5.0"
#addin "nuget:?package=Cake.Docker&version=0.9.6"
#addin "nuget:?package=dbup-core&version=4.1.0"
#addin "nuget:?package=dbup-sqlserver&version=4.1.0"
```

In my Docker file:

```
FROM microsoft/dotnet:2.1.400-sdk-stretch AS build-env

WORKDIR /app
COPY ./ ./
ENV PATH="/root/.dotnet/tools:${PATH}"
RUN dotnet tool install -g Cake.Tool --version 0.30.0
RUN dotnet cake build.cake
```

### Output Log

~~~sh
root@dc8cbcdce3b2:/app# dotnet cake build.cake -v=Diagnostic
Module directory does not exist.
NuGet.config not found.
Analyzing build script...
Analyzing /app/build.cake...
Processing build script...
Installing tools...
Retrieving package 'ReportGenerator 4.0.0-rc4' from '/root/.nuget/packages/'.
Package ReportGenerator.4.0.0-rc4 has already been installed.
Successfully installed 'ReportGenerator 4.0.0-rc4' to /app/tools
Executing nuget actions took 68.73 ms
Installing addins...
Found package 'System.Data.SqlClient 4.5.0' in '/app/tools/Addins'.
Package System.Data.SqlClient.4.5.0 has already been installed.
Successfully installed 'System.Data.SqlClient 4.5.0' to /app/tools/Addins
Executing nuget actions took 2.18 ms
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
The addin System.Data.SqlClient will reference System.Data.SqlClient.dll.
Found package 'Cake.Docker 0.9.6' in '/app/tools/Addins'.
Package Cake.Docker.0.9.6 has already been installed.
Successfully installed 'Cake.Docker 0.9.6' to /app/tools/Addins
Executing nuget actions took 0.46 ms
The addin Cake.Docker will reference Cake.Docker.dll.
Found package 'dbup-core 4.1.0' in '/app/tools/Addins'.
Package dbup-core.4.1.0 has already been installed.
Successfully installed 'dbup-core 4.1.0' to /app/tools/Addins
Executing nuget actions took 0.65 ms
The addin dbup-core will reference dbup-core.dll.
Found package 'dbup-sqlserver 4.1.0' in '/app/tools/Addins'.
Package dbup-sqlserver.4.1.0 has already been installed.
Successfully installed 'dbup-sqlserver 4.1.0' to /app/tools/Addins
Executing nuget actions took 0.57 ms
The addin dbup-sqlserver will reference dbup-sqlserver.dll.
Verifying assembly 'System.Data.SqlClient, Version=4.4.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
Error: System.IO.FileLoadException: Assembly with same name is already loaded
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromPath(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at Cake.Core.Reflection.AssemblyLoader.Load(FilePath path, Boolean verify) in C:\projects\cake\src\Cake.Core\Reflection\AssemblyLoader.cs:line 31
   at Cake.Core.Scripting.ScriptRunner.Run(IScriptHost host, FilePath scriptPath, IDictionary`2 arguments) in C:\projects\cake\src\Cake.Core\Scripting\ScriptRunner.cs:line 171
   at Cake.Commands.BuildCommand.Execute(CakeOptions options) in C:\projects\cake\src\Cake\Commands\BuildCommand.cs:line 41
   at Cake.CakeApplication.Run(CakeOptions options) in C:\projects\cake\src\Cake\CakeApplication.cs:line 45
   at Cake.Program.Main() in C:\projects\cake\src\Cake\Program.cs:line 73
~~~

Hi. Do we have an expected release for this? I've revisited the problem and I seem to hit the same problems...

If I do this:

Task("Default")
   .Does(() =>
{
    var connection = new SqlConnection("connectionString");
});

RunTarget("Default");

I get:
error CS0246: The type or namespace name 'SqlConnection' could not be found (are you missing a using directive or an assembly reference?)

If I do this:
#addin "System.Data.SqlClient&version=4.6.1"

Task("Default")
   .Does(() =>
{
    var connection = new SqlConnection("connectionString");
});

RunTarget("Default");

I get this:
Error: Assembly with same name is already loaded

If I do this:
 #addin System.Data.SqlClient&version=4.6.0&exclude=runtimes/**/*.dll

Task("Default")
   .Does(() =>
{
    var connection = new SqlConnection("connectionString");
});

RunTarget("Default");

I get this:
The type or namespace name 'SqlConnection' could not be found (are you missing a using directive or an assembly reference?)

This sounds promising but if I add a using statement:

#addin System.Data.SqlClient&version=4.6.0&exclude=runtimes/**/*.dll

using System.Data.SqlClient;

Task("Default")
   .Does(() =>
{
    var connection = new SqlConnection("connectionString");
});

RunTarget("Default");

I get this:
Resolving assembly System.Data.SqlClient.resources, Version=4.5.0.0, Culture=en-GB, PublicKeyToken=b03f5f7f11d50a3a
Exception while resolving assembly System.Data.SqlClient.resources: Could not load file or assembly 'System.Data.SqlClient.resources, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
One or more errors occurred. (System.Data.SqlClient is not supported on this platform.)

Is there a workaround or a recommended way of called Sql from Cake Core CLR?

(I'm on Windows)

--- a/src/Cake.Core/Polyfill/AssemblyHelper.cs
+++ b/src/Cake.Core/Polyfill/AssemblyHelper.cs
@@ -4,6 +4,7 @@
 
 using System;
 using System.Reflection;
+using System.Runtime.InteropServices;
 using Cake.Core.IO;
 #if NETCORE
 using Cake.Core.Reflection;
@@ -15,11 +16,7 @@ namespace Cake.Core.Polyfill
     {
         public static Assembly GetExecutingAssembly()
         {
-#if NETCORE
-            return typeof(CakeEnvironment).GetTypeInfo().Assembly;
-#else
             return Assembly.GetExecutingAssembly();
-#endif
         }
 
         public static Assembly LoadAssembly(AssemblyName assemblyName)
@@ -46,7 +43,40 @@ namespace Cake.Core.Polyfill
 
             // Make the path absolute.
             path = path.MakeAbsolute(environment);
-            return Assembly.LoadFrom(path.FullPath);
+
+            try
+            {
+                if (fileSystem.GetFile(path).IsClrAssembly())
+                {
+                    return Assembly.LoadFrom(path.FullPath);
+                }
+
+                if (environment.Platform.IsUnix())
+                {
+                    LoadUnixLibrary(path.FullPath, RTLD_NOW);
+                }
+                else
+                {
+                    LoadWindowsLibrary(path.FullPath);
+                }
+
+                return null;
+            }
+            catch (System.IO.FileLoadException)
+            {
+                // TODO: LOG
+                return null;
+            }
         }
+
+#pragma warning disable SA1310 // Field names should not contain underscore
+        private const int RTLD_NOW = 0x002;
+#pragma warning restore SA1310 // Field names should not contain underscore
+
+        [DllImport("libdl", EntryPoint = "dlopen")]
+        private static extern IntPtr LoadUnixLibrary(string path, int flags);
+
+        [DllImport("kernel32", EntryPoint = "LoadLibrary")]
+        private static extern IntPtr LoadWindowsLibrary(string path);
     }
 }

--- a/src/Cake.Core/Reflection/AssemblyLoader.cs
+++ b/src/Cake.Core/Reflection/AssemblyLoader.cs
@@ -29,7 +29,7 @@ namespace Cake.Core.Reflection
         public Assembly Load(FilePath path, bool verify)
         {
             var assembly = AssemblyHelper.LoadAssembly(_environment, _fileSystem, path);
-            if (verify)
+            if (verify && assembly != null)
             {
                 _verifier.Verify(assembly);
             }

--- a/src/Cake.Core/Scripting/ScriptRunner.cs
+++ b/src/Cake.Core/Scripting/ScriptRunner.cs
@@ -169,7 +169,11 @@ namespace Cake.Core.Scripting
                 if (host.Context.FileSystem.Exist(referencePath))
                 {
                     var assembly = _assemblyLoader.Load(referencePath, true);
-                    assemblies.Add(assembly);
+
+                    if (assembly != null)
+                    {
+                        assemblies.Add(assembly);
+                    }
                 }
                 else
                 {

--- a/src/Cake.NuGet.Tests/Fixtures/NuGetContentResolverFixture.cs
+++ b/src/Cake.NuGet.Tests/Fixtures/NuGetContentResolverFixture.cs
@@ -28,6 +28,7 @@ namespace Cake.NuGet.Tests.Fixtures
             Environment = FakeEnvironment.CreateUnixEnvironment();
             Environment.Runtime.BuiltFramework = new FrameworkName(framework);
             Environment.Runtime.Runtime = runtime;
+            Environment.Runtime.IsCoreClr = runtime == Runtime.CoreClr;
 
             FileSystem = new FakeFileSystem(Environment);
             Globber = new Globber(FileSystem, Environment);

--- a/src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs
+++ b/src/Cake.NuGet.Tests/Unit/NuGetContentResolverTests.cs
@@ -7,6 +7,7 @@ using Cake.Core.Diagnostics;
 using Cake.Core.Polyfill;
 using Cake.NuGet.Tests.Fixtures;
 using Cake.Testing;
+using Cake.Testing.Xunit;
 using NSubstitute;
 using Xunit;
 
@@ -227,7 +228,7 @@ namespace Cake.NuGet.Tests.Unit
                 Assert.Equal("/Working/lib/netstandard1.6/file.dll", result.ElementAt(0).Path.FullPath);
             }
 
-            [Theory]
+            [Theory(Skip = "Not possible to return files from root anymore.")]
             [InlineData(".NETStandard,Version=v1.6", Runtime.CoreClr)]
             [InlineData(".NETFramework,Version=v4.6.1", Runtime.Clr)]
             public void Should_Return_Files_When_Located_In_Root(string framework, Runtime runtime)
@@ -270,7 +271,7 @@ namespace Cake.NuGet.Tests.Unit
                 Assert.Equal($"/Working/lib/{expected}/file.dll", result.ElementAt(0).Path.FullPath);
             }
 
-            [Theory]
+            [Theory(Skip = "Not possible to return files from root anymore.")]
             [InlineData(".NETStandard,Version=v1.6", Runtime.CoreClr)]
             [InlineData(".NETFramework,Version=v4.6", Runtime.Clr)]
             public void Should_Return_From_Root_If_No_Compatible_Framework_Found(string framework, Runtime runtime)
@@ -289,7 +290,7 @@ namespace Cake.NuGet.Tests.Unit
                 Assert.Equal($"/Working/file.dll", result.ElementAt(0).Path.FullPath);
             }
 
-            [Theory]
+            [Theory(Skip = "Not possible to return files from root anymore.")]
             [InlineData(".NETStandard,Version=v1.6", Runtime.CoreClr)]
             [InlineData(".NETFramework,Version=v4.6.1", Runtime.Clr)]
             public void Should_Log_Warning_For_Files_Located_In_Root(string framework, Runtime runtime)
@@ -323,6 +324,65 @@ namespace Cake.NuGet.Tests.Unit
 
                 fixture.CreateCLRAssembly("/Working/ref/netstandard1.6/file.dll");
                 fixture.CreateCLRAssembly("/Working/ref/net46/file.dll");
+                fixture.FileSystem.CreateFile("/Working/lib/netstandard1.6/_._");
+                fixture.FileSystem.CreateFile("/Working/lib/net46/_._");
+
+                // When
+                var result = fixture.GetFiles();
+
+                // Then
+                Assert.Equal(0, result.Count);
+            }
+
+            [RuntimeFact(TestRuntime.CoreClr)]
+            public void Should_Return_Runtimes_Assemblies_If_CoreCLR()
+            {
+                // Given
+                var framework = ".NETCoreApp,Version=v2.0";
+                var runtime = Runtime.CoreClr;
+                var fixture = new NuGetAddinContentResolverFixture(framework, runtime);
+
+                fixture.CreateCLRAssembly("/Working/runtimes/win/lib/netstandard1.6/file.dll");
+                fixture.CreateCLRAssembly("/Working/runtimes/unix/lib/netstandard1.6/file.dll");
+                fixture.FileSystem.CreateFile("/Working/lib/netstandard1.6/_._");
+                fixture.FileSystem.CreateFile("/Working/lib/net46/_._");
+
+                // When
+                var result = fixture.GetFiles();
+
+                // Then
+                Assert.Equal(1, result.Count);
+            }
+
+            [RuntimeFact(TestRuntime.CoreClr)]
+            public void Should_Return_Native_Runtimes_Assemblies_If_CoreCLR()
+            {
+                // Given
+                var framework = ".NETCoreApp,Version=v2.0";
+                var runtime = Runtime.CoreClr;
+                var fixture = new NuGetAddinContentResolverFixture(framework, runtime);
+
+                fixture.CreateCLRAssembly("/Working/runtimes/win/native/file.dll");
+                fixture.CreateCLRAssembly("/Working/runtimes/linux/native/file.so");
+                fixture.CreateCLRAssembly("/Working/runtimes/osx/native/file.dylib");
+
+                // When
+                var result = fixture.GetFiles();
+
+                // Then
+                Assert.Equal(1, result.Count);
+            }
+
+            [RuntimeFact(TestRuntime.Clr)]
+            public void Should_Not_Return_Runtimes_Assemblies_If_Clr()
+            {
+                // Given
+                var framework = ".NETFramework,Version=v4.6.1";
+                var runtime = Runtime.Clr;
+                var fixture = new NuGetAddinContentResolverFixture(framework, runtime);
+
+                fixture.CreateCLRAssembly("/Working/runtimes/win/lib/netstandard1.6/file.dll");
+                fixture.CreateCLRAssembly("/Working/runtimes/unix/lib/netstandard1.6/file.dll");
                 fixture.FileSystem.CreateFile("/Working/lib/netstandard1.6/_._");
                 fixture.FileSystem.CreateFile("/Working/lib/net46/_._");
 

--- a/src/Cake.NuGet/Cake.NuGet.csproj
+++ b/src/Cake.NuGet/Cake.NuGet.csproj
@@ -26,5 +26,16 @@
     <PackageReference Include="NuGet.Resolver" Version="5.3.0" />
     <PackageReference Include="NuGet.Common" Version="5.3.0" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
+    <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="3.0.0" />
   </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <PackageReference Include="Microsoft.NETCore.Platforms" Version="2.2.2" GeneratePathProperty="true">
+      <PrivateAssets>All</PrivateAssets>
+    </PackageReference>
+    <EmbeddedResource  Include="$(PkgMicrosoft_NETCore_Platforms)\runtime.json">
+      <Link>runtime.json</Link>
+    </EmbeddedResource>
+  </ItemGroup>
+
 </Project>

--- a/src/Cake.NuGet/Cake.NuGet.csproj
+++ b/src/Cake.NuGet/Cake.NuGet.csproj
@@ -19,21 +19,21 @@
   </ItemGroup>
   <!-- Global packages -->
   <ItemGroup>
-    <PackageReference Include="NuGet.Frameworks" Version="5.3.0" />
-    <PackageReference Include="NuGet.Versioning" Version="5.3.0" />
-    <PackageReference Include="NuGet.Protocol" Version="5.3.0" />
-    <PackageReference Include="NuGet.Packaging" Version="5.3.0" />
-    <PackageReference Include="NuGet.Resolver" Version="5.3.0" />
-    <PackageReference Include="NuGet.Common" Version="5.3.0" />
+    <PackageReference Include="NuGet.Frameworks" Version="5.4.0" />
+    <PackageReference Include="NuGet.Versioning" Version="5.4.0" />
+    <PackageReference Include="NuGet.Protocol" Version="5.4.0" />
+    <PackageReference Include="NuGet.Packaging" Version="5.4.0" />
+    <PackageReference Include="NuGet.Resolver" Version="5.4.0" />
+    <PackageReference Include="NuGet.Common" Version="5.4.0" />
     <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
-    <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="3.0.0" />
+    <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="3.1.0" />
   </ItemGroup>
 
   <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
-    <PackageReference Include="Microsoft.NETCore.Platforms" Version="2.2.2" GeneratePathProperty="true">
+    <PackageReference Include="Microsoft.NETCore.Platforms" Version="3.1.0" GeneratePathProperty="true">
       <PrivateAssets>All</PrivateAssets>
     </PackageReference>
-    <EmbeddedResource  Include="$(PkgMicrosoft_NETCore_Platforms)\runtime.json">
+    <EmbeddedResource Include="$(PkgMicrosoft_NETCore_Platforms)\runtime.json">
       <Link>runtime.json</Link>
     </EmbeddedResource>
   </ItemGroup>

--- a/src/Cake.NuGet/NuGetContentResolver.cs
+++ b/src/Cake.NuGet/NuGetContentResolver.cs
@@ -4,13 +4,15 @@
 
 using System;
 using System.Collections.Generic;
-using System.Collections.ObjectModel;
 using System.Linq;
 using Cake.Core;
 using Cake.Core.Diagnostics;
 using Cake.Core.IO;
 using Cake.Core.Packaging;
+using NuGet.Client;
+using NuGet.ContentModel;
 using NuGet.Frameworks;
+using NuGet.RuntimeModel;
 
 namespace Cake.NuGet
 {
@@ -20,6 +22,19 @@ namespace Cake.NuGet
         private readonly ICakeEnvironment _environment;
         private readonly IGlobber _globber;
         private readonly ICakeLog _log;
+
+        private static readonly Lazy<RuntimeGraph> RuntimeGraph = new Lazy<RuntimeGraph>(() =>
+        {
+#if NETCORE
+            var assembly = typeof(NuGetContentResolver).Assembly;
+            using (var stream = assembly.GetManifestResourceStream($"{assembly.GetName().Name}.runtime.json"))
+            {
+                return JsonRuntimeFormat.ReadRuntimeGraph(stream);
+            }
+#else
+            return global::NuGet.RuntimeModel.RuntimeGraph.Empty;
+#endif
+        });
 
         public NuGetContentResolver(
             IFileSystem fileSystem,
@@ -56,82 +71,39 @@ namespace Cake.NuGet
         {
             if (!_fileSystem.Exist(path))
             {
-                return new List<IFile>();
+                return Array.Empty<IFile>();
             }
 
             // Get current framework.
-            var provider = DefaultFrameworkNameProvider.Instance;
-            var current = NuGetFramework.Parse(_environment.Runtime.BuiltFramework.FullName, provider);
+            var tfm = NuGetFramework.Parse(_environment.Runtime.BuiltFramework.FullName, DefaultFrameworkNameProvider.Instance);
 
-            // Get all ref assemblies.
-            var refAssemblies = _globber.GetFiles(path.FullPath + "/ref/**/*.dll");
+            // Get current runtime identifier.
+            string rid = _environment.Runtime.IsCoreClr ? Microsoft.DotNet.PlatformAbstractions.RuntimeEnvironment.GetRuntimeIdentifier() : null;
 
             // Get all candidate files.
             var pathComparer = PathComparer.Default;
-            var assemblies = GetFiles(path, package, new[] { path.FullPath + "/**/*.dll" })
-                .Where(file => !"Cake.Core.dll".Equals(file.Path.GetFilename().FullPath, StringComparison.OrdinalIgnoreCase)
-                               && file.IsClrAssembly()
-                               && !refAssemblies.Contains(file.Path, pathComparer))
-                .ToList();
+            var assemblies = GetFiles(path, package, new[] { path.FullPath + "/**/*.{dll,so,dylib}" })
+                .Where(file => !"Cake.Core.dll".Equals(file.Path.GetFilename().FullPath, StringComparison.OrdinalIgnoreCase))
+                .ToDictionary(x => path.GetRelativePath(x.Path).FullPath);
 
-            // Iterate all found files.
-            var comparer = new NuGetFrameworkFullComparer();
-            var mapping = new Dictionary<NuGetFramework, List<FilePath>>(comparer);
-            foreach (var assembly in assemblies)
+            var conventions = new ManagedCodeConventions(RuntimeGraph.Value);
+
+            var collection = new ContentItemCollection();
+            collection.Load(assemblies.Keys);
+
+            var criteria = conventions.Criteria.ForFrameworkAndRuntime(tfm, rid);
+            var managedAssemblies = collection.FindBestItemGroup(criteria, conventions.Patterns.RuntimeAssemblies);
+
+            var files = managedAssemblies?.Items.Select(x => assemblies[x.Path]).Where(x => x.IsClrAssembly()) ?? Enumerable.Empty<IFile>();
+
+            if (_environment.Runtime.IsCoreClr)
             {
-                // Get relative path.
-                var relative = path.GetRelativePath(assembly.Path);
-                var framework = ParseFromDirectoryPath(current, relative.GetDirectory());
-                if (!mapping.ContainsKey(framework))
-                {
-                    mapping.Add(framework, new List<FilePath>());
-                }
-                mapping[framework].Add(assembly.Path);
+                var nativeAssemblies = collection.FindBestItemGroup(criteria, conventions.Patterns.NativeLibraries);
+
+                files = (nativeAssemblies?.Items.Select(x => assemblies[x.Path]) ?? Enumerable.Empty<IFile>()).Concat(files);
             }
 
-            // Reduce found frameworks to the closest one.
-            var reducer = new FrameworkReducer();
-            var nearest = reducer.GetNearest(current, mapping.Keys);
-            if (nearest == null || !mapping.ContainsKey(nearest))
-            {
-                return new List<IFile>();
-            }
-
-            if (nearest == NuGetFramework.AnyFramework)
-            {
-                var framework = _environment.Runtime.BuiltFramework;
-                _log.Warning("Could not find any assemblies compatible with {0} in NuGet package {1}. " +
-                             "Falling back to using root folder of NuGet package.", framework.FullName, package.Package);
-            }
-
-            // Return the result.
-            return mapping[nearest].Select(p => _fileSystem.GetFile(p)).ToList();
-        }
-
-        private NuGetFramework ParseFromDirectoryPath(NuGetFramework current, DirectoryPath path)
-        {
-            var segments = path.Segments;
-
-            if (segments.Length == 1 &&
-                segments[0].Equals("lib", StringComparison.OrdinalIgnoreCase))
-            {
-                // Treat as AnyFramework if lib folder
-                return NuGetFramework.AnyFramework;
-            }
-
-            var queue = new Queue<string>(segments);
-            while (queue.Count > 0)
-            {
-                var other = NuGetFramework.Parse(queue.Dequeue(), DefaultFrameworkNameProvider.Instance);
-                var compatible = DefaultCompatibilityProvider.Instance.IsCompatible(other, current);
-                if (compatible || queue.Count == 0)
-                {
-                    return other;
-                }
-            }
-
-            // Treat as AnyFramework if root folder
-            return NuGetFramework.AnyFramework;
+            return files.ToArray();
         }
 
         private IReadOnlyCollection<IFile> GetToolFiles(DirectoryPath path, PackageReference package)

